### PR TITLE
feature(js): elgg/Ajax users get more access to underlying resources

### LINKED ISSUE #9767: Plugins using elgg/Ajax can't tell if action is successful
Old API used to sniff for errors and set `status` to `-1` if it found registered errors. Current `elgg/Ajax` implementation strips error messages from the response and expects actions to implement their own status logic hence there is no reliable way of telling if the action has succeeded.

### PR DESCRIPTION
**Pull request by mrclay**
All success/fail/complete callbacks and Promise handlers now receive the standard jQuery arguments, including `jqXHR`, `textStatus`, and `errorThrown`.

In successful responses, `jqXHR.AjaxData` contains the wrapper object passed through the `ajax_response` hooks, and `jqXHR.AjaxData.status` is set to 0 or -1 (indicating `register_error()` was called on the server).

The default error handler is improved so that server-encoded messages (sent from `register_error`/`system_message`) are displayed even on HTTP error responses. Note the API to send this type of responses has not yet been implemented, but is planned.

Updates ElggAjaxTest to use jquery-mockjax for better coverage and tests new features.

Makes developer/ajax_demo a little more like an integration test.

Fixes #9767

### CODE REVIEW

**Comment by mrclay** [comment]
@hypeJunction this is (hopefully) the less controversial bit of #9776. It doesn't "fix" #9767, but lays the groundwork for handling errors more cleanly.

File: views/default/elgg/Ajax.js [hunk 1]
```diff
@@ -33,11 +35,13 @@ define(function (require) {
...
-		 *     error : {Function} Error handler. Default is elgg.ajax.handleAjaxError. To cancel this altogether,
-		 *                        pass in function(){}.
-		 *     data  : {Object} Data to send to the server (optional). If set to a string (e.g. $.serialize)
-		 *                      then the request hook will not be called.
+		 *     url   : {String}   Path of the Ajax API endpoint (required)
+		 *     data  : {Object}   Data to send to the server (optional). If set to a string (e.g. $.serialize)
+		 *                        then the request hook will not be called.
+		 *     error : {Function} Error handler. With API version 2 this is only called on rare errors like
+		 *                        404s, missing views, or an exception. With API 3, most responses that use
+		 *                        register_error() also send HTTP error codes. The default handler is usually
```
Comment (hypeJunction) [inline views/default/elgg/Ajax.js]: Not yet

File: views/default/elgg/Ajax.js [hunk 2]
```diff
@@ -106,38 +129,77 @@ define(function (require) {
...
+				extract_metadata(data);
+
+				var params = {
+					options: orig_options
+				};
+				if (hook_type) {
+					data = elgg.trigger_hook(Ajax.RESPONSE_DATA_HOOK, hook_type, params, data);
+				}
+
+				jqXHR.ajaxData = data;
```
Comment (hypeJunction) [inline views/default/elgg/Ajax.js]: Why not add `jqXHR.errorDisplayed` and consider it a fix for https://github.com/Elgg/Elgg/issues/9767.  We just need a temporary solution to determine if error was shown.

**Comment by hypeJunction** [comment]
I like this

File: views/default/elgg/Ajax.js (hunk 1 above)
Comment (mrclay) [inline views/default/elgg/Ajax.js]: Forgot to rewrite this for this PR

**Comment by mrclay** [comment]
@Elgg/core any big objections to this? @hypeJunction is slowly building an [important PR](https://github.com/Elgg/Elgg/pull/9779) to bring testability to actions, pages, and routing, and this would help there.

**Comment by hypeJunction** [comment]
Would it be worth to branch off 2.x for this massive milestone?

**Comment by mrclay** [comment]
I pulled the trigger on this; it was blocking a feature I'm trying to work into elgg/Ajax

**Pull request merged**